scene/resources/video_stream.h

```diff
@@ -39,40 +40,103 @@ class VideoStreamPlayback : public Resource {
 public:
 	typedef int (*AudioMixCallback)(void *p_udata, const float *p_data, int p_frames);
 
-	virtual void stop() = 0;
-	virtual void play() = 0;
+protected:
+	String file_name;
+	Ref<FileAccess> file = nullptr;
+
+	AudioMixCallback mix_callback = nullptr;
+	void *mix_udata = nullptr;
+
+	static void _bind_methods();
+	GDVIRTUAL0(_stop);
+	GDVIRTUAL0(_play);
+	GDVIRTUAL0RC(bool, _is_playing);
+	GDVIRTUAL1(_set_paused, bool);
+	GDVIRTUAL0RC(bool, _is_paused);
+	GDVIRTUAL1(_set_loop, bool);
+	GDVIRTUAL0RC(bool, _has_loop);
+	GDVIRTUAL0RC(double, _get_length);
+	GDVIRTUAL0RC(double, _get_playback_position);
+	GDVIRTUAL1(_seek, double);
+	GDVIRTUAL1(_set_audio_track, int);
+	GDVIRTUAL0RC(Ref<Texture2D>, _get_texture);
+	GDVIRTUAL1(_update, double);
+	GDVIRTUAL0RC(int, _get_channels);
+	GDVIRTUAL0RC(int, _get_mix_rate);
+
+	GDVIRTUAL0R(bool, _file_opened);
+
+	GDVIRTUAL0(_initialize);
+	GDVIRTUAL0(_cleanup);
+
+	PackedByteArray file_read(uint64_t length);
+	void file_seek(int pos);
+	uint64_t file_pos();
+	uint64_t file_len();
+	int mix_audio(int num_frames, PackedFloat32Array buffer = {}, int offset = 0);
 
-	virtual bool is_playing() const = 0;
+public:
+	VideoStreamPlayback();
+	virtual ~VideoStreamPlayback();
+
+	bool open_file(const String &p_file);
+
+	virtual void stop();
+	virtual void play();
 
-	virtual void set_paused(bool p_paused) = 0;
-	virtual bool is_paused() const = 0;
+	virtual bool is_playing() const;
 
-	virtual void set_loop(bool p_enable) = 0;
-	virtual bool has_loop() const = 0;
+	virtual void set_paused(bool p_paused);
+	virtual bool is_paused() const;
 
-	virtual double get_length() const = 0;
+	virtual void set_loop(bool p_enable);
+	virtual bool has_loop() const;
 
-	virtual double get_playback_position() const = 0;
-	virtual void seek(double p_time) = 0;
+	virtual double get_length() const;
 
-	virtual void set_audio_track(int p_idx) = 0;
+	virtual double get_playback_position() const;
+	virtual void seek(double p_time);
 
-	virtual Ref<Texture2D> get_texture() const = 0;
+	virtual void set_audio_track(int p_idx);
 
-	virtual void update(double p_delta) = 0;
+	virtual Ref<Texture2D> get_texture() const;
+	virtual void update(double p_delta);
 
-	virtual void set_mix_callback(AudioMixCallback p_callback, void *p_userdata) = 0;
-	virtual int get_channels() const = 0;
-	virtual int get_mix_rate() const = 0;
+	virtual void set_mix_callback(AudioMixCallback p_callback, void *p_userdata);
+	virtual int get_channels() const;
+	virtual int get_mix_rate() const;
 };
 
 class VideoStream : public Resource {
 	GDCLASS(VideoStream, Resource);
-	OBJ_SAVE_TYPE(VideoStream); // Saves derived classes with common type so they can be interchanged.
+	OBJ_SAVE_TYPE(VideoStream);
+
+protected:
+	static void
+	_bind_methods();
+
+	// Using Object* to bypass godot-cpp$652
+	GDVIRTUAL0R(Object *, _instantiate_playback);
```

Comment:
This should return Ref<VideoStreamPlayback>. I'll fix it.

Reply:
(Since https://github.com/godotengine/godot-cpp/issues/652 was fixed in december)